add contributing doc: how to setup env + write new package (#135)
wolfi-dev/os: CONTRIBUTING.md

# Contributing to wolfi-dev/os

> * [Setup development environmment](#setup-development-environment)

Reviewer: Complete the TOC here, you can use https://github.com/kubernetes-sigs/mdtoc to generate it automatically.

## Setup development environment

> To ease the development of Wofli OS, you can use `chainguard-images/sdk` container image that already include both [apko](https://github.com/chainguard-dev/apko) & [melange](https://github.com/chainguard-dev/melange).

Reviewer: ```suggestion
To ease the development of Wolfi OS, you can use `chainguard-images/sdk` container image that already includes both [apko](https://github.com/chainguard-dev/apko) & [melange](https://github.com/chainguard-dev/melange).
```

To start a development environment, do

```sh
make dev-container
```

What it does is start the `chainguard-images/sdk` image and mount the current working directory into it.

> Now, the `Makefile` script assume that there're melange folder & melange binary in there. I change it locally to the following because it's where they are in `chainguard-image/sdk`.

Reviewer: Where is _in there_?  Try to rephrase this paragraph. I don't fully get what is happening here, but it sounds like a bug we should fix in the scripts.

Author: when i use `chainguard-imgaes/sdk`, the melange dir is `/usr/share/melange` and not `../melange` as in the Makefile. that's what i meant.

```
MELANGE_DIR ?= /usr/share/melange
MELANGE ?= $(shell which melange)
```

## Write your first Wolfi package

> Wofie is using melange to build packages. If you want to learn the details what commands are used when building packages, they are in the [`ci-build.yaml`](.github/workflows/ci-build.yaml) and [Makefile](Makefile).

Reviewer: I think this will change at some point, but I reworked the phrase a little bit:

```suggestion
Wolfi packages are built using melange. If you want to learn how packages are built, you can see all the details in the [`ci-build`](.github/workflows/ci-build.yaml) workflow and in the [Makefile](Makefile).
```

> Start by creating a YAML file named `<your-package-name>.yaml` in the repo's root directory. If you have patches, create a folder with same name and put patches there.

Reviewer: ```suggestion
Start by cloning this repository and create a YAML file named `<your-package-name>.yaml` in its root directory. If you have any patches, create a folder with the same name and put them there.
```

> Add a new entry for your package in Makefile like this

Reviewer: ```suggestion
Add a new entry for your package in the Makefile like this
```

```
$(eval $(call build-package,<your-package-name>,<version>-r<epoch>))
```

> Once you're done writing the new package, you can test building it with `make packages/<your-package-name>`.

Reviewer: ```suggestion
Once you're done writing the new package configuration file, you can test it by triggering a build with `make packages/<your-package-name>`.
```

## Some tips

> - melange has a few pipelines included, ready to be used. You can see the the list & source code of the builtin pipelines [here in melange repo](https://github.com/chainguard-dev/melange/tree/main/pkg/build/pipelines).

Reviewer: ```suggestion
- melange has a few built-in pipelines. You can see their source code [in the melange repository](https://github.com/chainguard-dev/melange/tree/main/pkg/build/pipelines).
```

> - Make sure to pump the epoch if you're changing the package's build script without updating the version. Also reset the epoch back to 0 when you update the package version.

Reviewer: More than a tip, bumping versions and epoch handling should be a section by itself. Move it out so the tip section to a **Package Versioning** section.

- You don't need to add `environment.contents.repositories` and `environment.contents.keyring`. Those are added automatically in the `ci-build.yaml` script during CI.

- For patching CVEs, you can follow the [documentation here](HOW_TO_PATCH_CVES.md).

> - If you don't want to build all the packages locally, you can install `gsutil` and use it to sync the prebuilt packages from `gs://wolfi-production-registry-destination/os/` bucket: 

Reviewer: Why would I need to build all packages? Add more context here to help folks understand what happens when you don't have the apks locally.

Author: what's the best practice here? what do you normally do?

Reviewer 2: @tuananh Just so I understand, what are you trying to accomplish by having all of the packages locally?

<!-- toc -->
- [Setup development environment](#setup-development-environment)
- [Write your first Wolfi package](#write-your-first-wolfi-package)
- [Package versioning](#package-versioning)
- [Some tips](#some-tips)
<!-- /toc -->

## Setup development environment

To ease the development of Wolfi OS, you can use `chainguard-images/sdk` container image that already includes both [apko](https://github.com/chainguard-dev/apko) & [melange](https://github.com/chainguard-dev/melange).

To start a development environment, do

```sh
make dev-container
```

What it does is start the `chainguard-images/sdk` image and mount the current working directory into it.

Now, the `Makefile` script assume that there're melange folder & melange binary the upper level dir (`../melange`). I change it locally to the following because it's where they are in `chainguard-image/sdk`.

```
MELANGE_DIR ?= /usr/share/melange
MELANGE ?= $(shell which melange)
```

## Write your first Wolfi package

Wolfi packages are built using melange. If you want to learn how packages are built, you can see all the details in the [`ci-build`](.github/workflows/ci-build.yaml) workflow and in the [Makefile](Makefile).

> Start by cloning this repository and create a YAML file named `<your-package-name>.yaml` in its root directory. If you have any patches, create a folder with the same name and put them there.

Reviewer 2: >Start by cloning this repository

We might want to put this "cloning" instruction in the first section, since that would be the first place in reading this doc that the repo needs to have been cloned. WDYT?

Add a new entry for your package in the Makefile like this

```
$(eval $(call build-package,<your-package-name>,<version>-r<epoch>))
```

Once you're done writing the new package configuration file, you can test it by triggering a build with `make packages/<your-package-name>`.

## Package versioning

> - When bumping version of a package, you will need to update the version, epoch & sha256 in package YAML file as well as Makefile.

Reviewer 2: The checksum isn't used in the Makefile. Also just a note, not all package YAML files use a sha256 value.

- `epoch` needs to be bump when package version remains the same but something else changes. `epoch` needs to be reset to 0 when it's a new version of the package.

> ## Some tips

Reviewer 3: FWIW it's slightly experimental, but maybe we can mention the `wolfictl lint` command in this section. (And maybe `./lint.sh`) So that contributors can run some checks just before submitting the PR. WDYT? (We're already running these in the CI: https://github.com/wolfi-dev/os/pull/280/files)

Reviewer 2: I like that idea
